libs/counterfactuals/src/lib/CounterfactualComponent.tsx

```diff
@@ -0,0 +1,295 @@
+    this.setState({ customPointLength });
+  };
+
+  // fetch prediction for temporary point
+  private fetchData = (fetchingReference: { [key: string]: any }): void => {
+    if (!this.props.invokeModel) {
+      return;
+    }
+    if (this.state.request !== undefined) {
+      this.state.request.abort();
+    }
+    const abortController = new AbortController();
+    const promise = getFetchPredictionPromise(
+      fetchingReference,
+      this.context.jointDataset,
+      this.props.invokeModel
+    );
+    fetchingReference[JointDataset.PredictedYLabel] = undefined;
+    this.setState(
+      { errorMessage: undefined, request: abortController },
+      async () => {
+        try {
+          const fetchedData = await promise;
+          // returns predicted probabilities
+          if (Array.isArray(fetchedData[0])) {
+            fetchingReference = getPredictedProbabilities(
+              fetchingReference,
+              fetchedData
+            );
+          } else {
+            // prediction is a scalar, no probabilities
+            fetchingReference[JointDataset.PredictedYLabel] = fetchedData[0];
+          }
+          if (this.context.jointDataset.hasTrueY) {
+            JointDataset.setErrorMetrics(
+              fetchingReference,
+              this.context.modelMetadata.modelType
+            );
+          }
+          this.setState({ request: undefined });
+        } catch (error) {
+          if (error.name === "AbortError") {
+            return;
+          }
+          if (error.name === "PythonError") {
+            this.setState({ errorMessage: error.message });
+          }
+        }
+      }
+    );
+  };
+
+  private setCustomRowProperty = (
+    key: string | number,
+    isString: boolean,
+    newValue?: string | number
+  ): void => {
+    if (!this.temporaryPoint || (!newValue && newValue !== 0)) {
+      return;
+    }
+    const editingData = this.temporaryPoint;
+    if (isString) {
+      editingData[key] = newValue;
+      this.forceUpdate();
+    } else {
+      const asNumber = +newValue;
+      // because " " evaluates to 0 in js
+      const isWhitespaceOnly = /^\s*$/.test(newValue?.toString());
+      if (Number.isNaN(asNumber) || isWhitespaceOnly) {
+        this.validationErrors[key] =
+          localization.Interpret.WhatIfTab.nonNumericValue;
+        this.forceUpdate();
+      } else {
+        editingData[key] = asNumber;
+        this.validationErrors[key] = undefined;
+        this.forceUpdate();
+        this.fetchData(editingData);
+      }
+    }
+  };
+
+  private setCustomRowPropertyComboBox = (
+    key: string | number,
+    index?: number,
+    value?: string
+  ): void => {
+    if (!this.temporaryPoint || (!value && !index)) {
+      return;
+    }
+    const editingData = this.temporaryPoint;
+    if (index !== undefined) {
+      // User selected/de-selected an existing option
+      editingData[key] = index;
+    }
+    this.forceUpdate();
```

Comment:
why is the forceUpdate needed before fetchData?  I see this above as well.

Reply:
This PR simply refactors the CounterfactualChart component without editing the detailed methods in the component, so we can make sure we don't block the logic while refactoring the component. The forceUpdate was there before.

But yeah it's worthwhile to look into how we can improve this by removing forceUpdate if it doesn't do anything here.
Created a task and I will look into this later: https://msdata.visualstudio.com/Vienna/_queries/edit/1923957/?triage=true
Thanks!!